fieldsolver/derivatives.cpp

```diff
@@ -485,3 +485,137 @@ void calculateBVOLDerivativesSimple(
 
    phiprof::stop("Calculate volume derivatives",N_cells,"Spatial Cells");
 }
+
+/*! \brief Returns volumetric E of cell
+ *
+ */
+std::array<Real, 3> getE(SpatialCell* cell)
+{
+   return std::array<Real, 3> { {cell->parameters[CellParams::EXVOL], cell->parameters[CellParams::EYVOL], cell->parameters[CellParams::EZVOL]} };
+}
+
+/*! \brief Returns perturbated volumetric B of cell
+ *
+ */
+std::array<Real, 3> getPerB(SpatialCell* cell)
+{
+   return std::array<Real, 3> { {cell->parameters[CellParams::PERBXVOL], cell->parameters[CellParams::PERBYVOL], cell->parameters[CellParams::PERBZVOL]} };
+}
+
+/*! \brief Calculates momentum density of cell
+ *
+ */
+std::array<Real, 3> getP(SpatialCell* cell)
+{
+   Real rho = cell->parameters[CellParams::RHOM];
+   return std::array<Real, 3> { {rho * cell->parameters[CellParams::VX], rho * cell->parameters[CellParams::VY], rho * cell->parameters[CellParams::VZ]} };
+}
+
+/*! \brief Calculates EM field energy for spatial cell with only perturbated magnetic field
+ *
+ */
+Real calculateU1(SpatialCell* cell)
+{
+   std::array<Real, 3> E = getE(cell);
+   std::array<Real, 3> B = getPerB(cell);
+   return 0.5 * (
+      physicalconstants::EPS_0 * (pow(E[0], 2) + pow(E[1], 2) + pow(E[2], 2)) + 
+      1.0/physicalconstants::MU_0 * (pow(B[0], 2) + pow(B[1], 2) + pow(B[2], 2))
+   );
+}
+
+/*! \brief Low-level scaled gradients calculation
+ * 
+ * For the SpatialCell* cell and its neighbors, calculate scaled gradients and their maximum alpha
```

Comment:
What are they scaled to though? Is it always dX/X?

Reply:
![kuva](https://user-images.githubusercontent.com/12125199/89992464-f9f05c80-dc8d-11ea-9982-009987ab09d9.png)

It's these guys from the GUMICS paper.

Reply:
Alright! A reference to that in the source code comment would be nice.